src/ctia/schemas/core.clj

```diff
@@ -38,6 +43,144 @@
    :created java.util.Date
    (s/optional-key :modified) java.util.Date})
 
+(s/defschema RealizeFnServices
+  "Maps of service functions available for realize-fns"
+  {:ConfigService {:get-in-config (s/=>* s/Any
+                                         [[s/Any]]
+                                         [[s/Any] s/Any])}
+   :StoreService {:read-store (s/pred ifn?)} ;;varags
+   :GraphQLNamedTypeRegistryService
+   {:get-or-update-named-type-registry
+    (s/=> graphql.schema.GraphQLType
+          s/Str
+          (s/=> graphql.schema.GraphQLType))}
+   :IEncryption {:encrypt (s/=> s/Any s/Any)
+                 :decrypt (s/=> s/Any s/Any)}})
+
+(s/defn APIHandlerServices->RealizeFnServices
+  :- RealizeFnServices
+  [services :- APIHandlerServices]
+  (select-keys-in
+    services
+    [:ConfigService [:get-in-config]]
+    [:StoreService [:read-store]]
+    [:GraphQLNamedTypeRegistryService [:get-or-update-named-type-registry]]
+    [:IEncryption [:decrypt :encrypt]]))
+
+(s/defschema GraphQLRuntimeOptions
+  "A map of options to resolve a DelayedGraphQLValue"
+  {:services RealizeFnServices})
+
+(defn delayed-graphql-value?
+  "A flat predicate deciding if the argument is delayed."
+  [v]
+  (fn? v))
```

Comment:
I wasn't comfortable using anything but a `clojure.lang.Fn` check here. A realize-fn seems to return just about anything except a `fn`.

Reply:
This look really dangerous. But considering the CTIA code is quite stable now this should be fine.

Another option would be to be extra protective and have a ns that deal with "delayed" objects.
A record, or simply `{:delayed f}`.
So the check will be a lot easier to really trust.
So instead of replacing all `(defn foo [& args] body)` by `(defn foo [& args] (fn [& args] body))`
you could instead return `(defn foo [& args] {:ctia.delayed/delayed (fn [& args] body))})`
so the `delayed-graphql-value?` would probably be a lot safer.

Reply:
I had the same thought overnight too. I think I'll do something like you suggest.

Reply:
I introduced a `ctia.graphql.delayed` namespace that encapsulates a simple wrapper datatype.